src/twisted/application/internet.py

```diff
@@ -523,13 +524,375 @@ def policy(attempt):
+                     outputs=[_notifyWaiters])
+    _connecting.upon(_connectionFailed, enter=_waiting,
+                     outputs=[_wait])
+
+    _waiting.upon(start, enter=_waiting,
+                  outputs=[])
+    _waiting.upon(stop, enter=_disconnecting,
+                  outputs=[_waitForStop, _stopRetrying],
+                  collector=firstResult)
+    _waiting.upon(_reconnect, enter=_connecting,
+                  outputs=[_connect])
+
+    _connected.upon(start, enter=_connected,
+                    outputs=[])
+    _connected.upon(stop, enter=_disconnecting,
+                    outputs=[_waitForStop, _disconnect],
+                    collector=firstResult)
+    _connected.upon(_clientDisconnected, enter=_waiting,
+                    outputs=[_forgetConnection, _wait])
+
+    _disconnecting.upon(start, enter=_restarting,
+                        outputs=[_resetFailedAttempts])
+    _disconnecting.upon(stop, enter=_disconnecting,
+                        outputs=[])
+    _disconnecting.upon(_clientDisconnected, enter=_stopped,
+                        outputs=[_cancelConnectWaiters,
+                                 _finishStopping,
+                                 _forgetConnection])
+    _disconnecting.upon(_connectionFailed, enter=_stopped,
+                        outputs=[_cancelConnectWaiters, _finishStopping])
+
+    _restarting.upon(start, enter=_restarting,
+                     outputs=[])
+    _restarting.upon(stop, enter=_disconnecting,
+                     outputs=[])
+    _restarting.upon(_clientDisconnected, enter=_connecting,
+                     outputs=[_finishStopping, _connect])
+    _restarting.upon(_connectionFailed, enter=_stopped,
+                     outputs=[_finishStopping, _connect])
+
+    _stopped.upon(start, enter=_connecting,
+                  outputs=[_connect])
+    _stopped.upon(stop, enter=_stopped,
+                  outputs=[])
+
+    _init.upon(whenConnected, enter=_init,
+               outputs=[_awaitingConnection],
+               collector=firstResult)
+    _connecting.upon(whenConnected, enter=_connecting,
+                     outputs=[_awaitingConnection],
+                     collector=firstResult)
+    _waiting.upon(whenConnected, enter=_waiting,
+                  outputs=[_awaitingConnection],
+                  collector=firstResult)
+    _connected.upon(whenConnected, enter=_connected,
+                    outputs=[_currentConnection],
+                    collector=firstResult)
+    _disconnecting.upon(whenConnected, enter=_disconnecting,
+                        outputs=[_awaitingConnection],
+                        collector=firstResult)
+    _restarting.upon(whenConnected, enter=_restarting,
+                     outputs=[_awaitingConnection],
+                     collector=firstResult)
+    _stopped.upon(whenConnected, enter=_stopped,
+                  outputs=[_noConnection],
+                  collector=firstResult)
+
+    del firstResult
```

Comment:
Maybe just define this as `_firstResult` outside the scope of the class so you don't need to `del` it?